src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java

```diff
@@ -1885,89 +1901,80 @@ private static void validateChapterWiseTesting(Iterator<Node> itrChecks,
                 assertWithMessage(styleName + "_style.xml rule '" + ruleName + "' module '"
-                        + moduleName + "' is missing the config link: config").fail();
+                        + moduleName + "' is missing the config link").fail();
             }
 
             final String configUrl = config.getAttributes().getNamedItem("href")
                     .getTextContent();
 
-            final String expectedUrl = "https://github.com/search?q="
-                    + "path%3Asrc%2Fmain%2Fresources%20path%3A**%2F" + styleName
+            final String wholeConfigUrl = partialConfigUrl
                     + "_checks.xml+repo%3Acheckstyle%2Fcheckstyle+" + moduleName;
 
             assertWithMessage("google_style.xml rule '" + ruleName + "' module '"
                     + moduleName + "' should have matching config url")
                     .that(configUrl)
-                    .isEqualTo(expectedUrl);
-
+                    .isEqualTo(wholeConfigUrl);
         }
 
-        if (itrConfigs.hasNext()) {
+        if (itrSample.hasNext()) {
             assertWithMessage(styleName + "_style.xml rule '" + ruleName + "' should have checks"
-                    + " if it has config/test links")
+                    + " if it has check/config links")
                     .that(hasChecks)
                     .isTrue();
 
-            final Node config = itrConfigs.next();
-            final String configUrl = config.getAttributes().getNamedItem("href")
+            final Node sample = itrSample.next();
+            final String inputFolderUrl = sample.getAttributes().getNamedItem("href")
                     .getTextContent();
-            final String extractedRuleName = getExtractedRuleName(ruleName);
+            final String extractedChapterNumber = getExtractedChapterNumber(ruleName);
+            final String extractedSectionNumber = getExtractedSectionNumber(ruleName);
 
             assertWithMessage("google_style.xml rule '" + ruleName + "' rule '"
-                    + "' should have matching test url")
-                    .that(configUrl)
+                    + "' should have matching sample url")
+                    .that(inputFolderUrl)
                     .startsWith("https://github.com/checkstyle/checkstyle/"
-                            + "blob/master/src/it/java/com/google"
-                            + "/checkstyle/test/");
+                            + "tree/master/src/it/resources/com/google/checkstyle/test/");
 
             assertWithMessage("google_style.xml rule '" + ruleName
-                    + "' should have matching test url")
-                    .that(configUrl)
-                    .endsWith("/" + extractedRuleName + "Test.java");
+                    + "' should have matching sample url")
+                    .that(inputFolderUrl)
+                    .containsMatch("/chapter" + extractedChapterNumber
+                            + ".*/rule" + extractedSectionNumber + ".*");
```

Comment:
`.*` is too broad, as it will match everything. Why not `[^/]+` for the first one.

I would need an example of what the 2nd one should be matching.

Reply:
second `.*` is used to match the section's folder name, for [3.3.1 No wildcard imports](https://checkstyle.org/google_style.html#a3.3.1)  the folder name is [rule331nowildcard](https://github.com/checkstyle/checkstyle/tree/master/src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule331nowildcard) so we need to match it

Reply:
I have changed regex a bit as per your suggestions:

https://github.com/checkstyle/checkstyle/blob/b7ee527133f73fde1dedb488edc17c231c3d2722/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1942-L1947

Reply:
I checked `containsMatch` is a find command. It is not matching the entire string. `inputFolderUrl` is the entire URL, this is why you don't have to match `https://`.

I don't see why we need the ending `.+$` and it can be removed completely. However, I feel it should be replaced with something like `[^\d\.]` (smilar to `[^/]+`) as we have some sections with numbers longer than others. We don't want to match another section by mistake. We should probably do something similar to chapter number for things like 1 and 10.

Reply:
> `inputFolderUrl` is the entire URL, this is why you don't have to match `https://`.

yes, we're validating `https://` and other starting content of `inputFolderUrl` here:

https://github.com/checkstyle/checkstyle/blob/cc4909e7907de21f6026e5c703cf9e4e128468a9/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1935-L1939

to ensure it is an input folder link.

To ensure that the provided folder link is of the current section, we're using the currently being reviewed test.

>  I feel it should be replaced with something like [^\d\.] (smilar to [^/]+) as we have some sections with numbers longer than others. We don't want to match another section by mistake. We should probably do something similar to chapter number for things like 1 and 10.

I have made changes according to your suggestions:

https://github.com/checkstyle/checkstyle/blob/cc4909e7907de21f6026e5c703cf9e4e128468a9/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1944-L1946

but Im not able to understand:

> as we have some sections with numbers longer than others. We don't want to match another section by mistake. We should probably do something similar to chapter number for things like 1 and 10.

if you don't mind, can you explain it in more details pls?

Reply:
@rnveach fyi I have changed the `[^\\d\\.]` regex a bit as [CI](https://app.circleci.com/pipelines/github/checkstyle/checkstyle/27855/workflows/9127bd50-075c-42a6-add3-2aca90c7d459/jobs/687720/artifacts) was complaining about it earlier:

https://github.com/checkstyle/checkstyle/blob/e30fc3659648bbcb930040dcfc1abd826316b00e/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1944-L1946

Reply:
> > as we have some sections with numbers longer than others. We don't want to match another section by mistake. We should probably do something similar to chapter number for things like 1 and 10.
> if you don't mind, can you explain it in more details pls?

Think of this as validation. You are validating what you tell it to, and you are telling it doesn't have to validate things when you use `.*`.

Your original validation is basically `/chapter{Number}.*`. This works fine if the user enters `/chapter1myname` on chapter 1, but what if the user accidentally writes `/chapter13myname` for the same chapter? Your validation still works even though it is confusingly looks like it should be for chapter 13. You want to validate as much as you reasonable can. If this is chapter 1, it should be a 1 with no more numbers after it.

`.*` is very powerful. You will want to look at its usages very carefully and see if there any possible issue.

Reply:
````
 "/chapter" + extractedChapterNumber 
           + "[^/]+/rule" + extractedSectionNumber + "[^\\d.]"); 
````

--->

````
 "/chapter" + extractedChapterNumber 
           + "[^\\d][^/]+/rule" + extractedSectionNumber + "[^\\d.]"); 
````

Reply:
done,

https://github.com/checkstyle/checkstyle/blob/664a71df19233e1e6444dc07ea9e01e564eab158/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1944-L1946

this ensures that there's no extra number after the required ones. 

Changed `[^\\d.]` to `\\D`, both of them were meeting our expectations